File: MR.cpp
#include <iostream>
#include <ctime>

#include "int1024.h"

/* Miller-Rabin test algorithm */
bool MR(const Int1024& N, int t = 10) {
    if (N < 5) return N == 2 || N == 3;
    // build mods regarding N
    Int1024::buildMods(N);
    // constants that may be often used
    Int1024 ONE(1), TWO(2);
    Int1024 N_MINUS_1 = N - 1;
    Int1024 N_MINUS_3 = N - 3;
    // decompose N - 1 into the form m * 2 ^ k
    Int1024 m; int k = N_MINUS_1.get2Power(m);
    // test for t rounds
    for (int i = 0; i < t; ++i) {
        // 2 <= a + 2 <= N - 2
        Int1024 a = N_MINUS_3.randomSmaller();
        // y = (a ^ m) % N
        Int1024 y = (a + 2).pow(m);
        if (y != ONE && y != N_MINUS_1) {
            for (int j = 1; j < k; ++j) {
                y = y.pow(TWO); // mod N included
                if (y == N_MINUS_1) break;
                if (y == ONE) return false;
            }
            if (y != N_MINUS_1) return false;
        }
    }
    return true;
}

int main() {
    double start = clock();
    // read the integers
    Int1024 N1("data/N1"), N2("data/N2");
    // apply M-R test
    std::cout << "M-R test for N1: " << (MR(N1) ? "Is prime" : "Not prime") << std::endl;
    std::cout << "M-R test for N2: " << (MR(N2) ? "Is prime" : "Not prime") << std::endl;
    std::cout << "Time usage (in seconds): " << (clock() - start) / CLOCKS_PER_SEC << std::endl;
    return 0;
}

File: MR.py
from random import seed, randint

def decompose(N):
    ''' Decompose N into the form m * 2 ^ k.
        :returns: (m, k).
    '''
    m, k = N, 0
    while m & 1 == 0:
        m, k = m >> 1, k + 1
    return m, k


def MR(N, t=10):
    ''' Miller-Rabin test implementation.
        :param N: The integer to be tested.
        :param t: Number of loops when testing.
        :returns: Whether N is (most likely) a prime.
    '''
    if N < 5 or not N & 1:
        return N in (2, 3)
    m, k = decompose(N - 1)
    for i in range(t):
        a = randint(2, N - 2)
        y = pow(a, m, mod=N)
        if y not in (1, N - 1):
            for j in range(k - 1):
                y = pow(y, 2, mod=N)
                if y == 1:
                    return False
                if y == N - 1:
                    break
            if y != N - 1:
                return False
    return True


# Read the integers
with open("data/N1", "rb") as f:
    N1 = int.from_bytes(f.read(), "big")
with open("data/N2", "rb") as f:
    N2 = int.from_bytes(f.read(), "big")
# print(hex(N1))
# print(hex(N2))


# Apply M-R test
seed(2018011368) # reproducibility
print("M-R prime test result for N1:", MR(N1))
print("M-R prime test result for N2:", MR(N2))


# There are 9592 primes in [1, 100000]
print('[t = 1] "prime" count in 100000:', sum(MR(i, 1) for i in range(100000)))
print('[t = 2] "prime" count in 100000:', sum(MR(i, 2) for i in range(100000)))
print('[t = 3] "prime" count in 100000:', sum(MR(i, 3) for i in range(100000)))
